File: src/main/java/fr/ynerant/leveleditor/game/BasicTower.java
package fr.ynerant.leveleditor.game;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class BasicTower extends Tower {
    public BasicTower(int x, int y) {
        super(x, y);
    }

    @Override
    public String getName() {
        return "basictower";
    }

    @Override
    public int getDamagePerShot() {
        return 1;
    }

    @Override
    public long getPeriod() {
        return 5;
    }

    @Override
    public int getPrice() {
        return 10;
    }

    @Override
    Collection<Mob> _filterDetectedMobs(Collection<Mob> mobs) {
        List<Mob> filtered = new ArrayList<>();

        for (Mob mob : mobs) {
            if (mob.getX() == getX() || mob.getY() == getY())
                filtered.add(mob);
        }

        return filtered;
    }
}

File: src/main/scala/fr/ynerant/leveleditor/game/towers/BasicTower.scala
package fr.ynerant.leveleditor.game.towers

import fr.ynerant.leveleditor.game.GameFrame

import scala.util.Random

class BasicTower(override val x: Int, override val y: Int) extends Tower(x, y) {
	override def getName = "basictower"

	override def getDamagePerShot = if (isUpgraded) 3 else 1

	override def getPeriod = 5

	override def getPrice = 10

	override private[towers] def _shot(game: GameFrame): Unit = {
		var l = game.getMobs.filter(mob => Math.pow(mob.getX - getX, 2) + Math.pow(mob.getY - getY, 2) <= 9).toList
		l = Random.shuffle(l)
		if (l.nonEmpty)
			l.head.hit(getDamagePerShot)
	}
}
